lib/cretonne/src/simple_gvn.rs

```diff
@@ -0,0 +1,133 @@
+//! A simple GVN pass.
+
+use flowgraph::{ControlFlowGraph};
+use dominator_tree::DominatorTree;
+use ir::{Cursor, DataFlowGraph, InstructionData, Function, Inst, Opcode};
+use std::hash::{Hash, Hasher};
+use std::collections::HashMap;
+
+/// Test whether the given opcode is unsafe to even consider for GVN.
+fn trivially_unsafe_for_gvn(opcode: Opcode) -> bool {
+    opcode.is_call() || opcode.is_branch() || opcode.is_terminator() ||
+    opcode.is_return() || opcode.can_trap()
+}
+
+#[derive(Debug)]
+struct VisibleInst(InstructionData);
+
+impl Hash for VisibleInst {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        // TODO: Is a better way to do this? Should we just make InstructionData
+        // implement Hash itself?
+        use ir::instructions::InstructionData::*;
+        match self.0 {
+            Nullary { .. } => {},
+            Unary { arg, .. } => arg.hash(state),
+            Binary { args, .. } => args.hash(state),
+            Ternary { args, .. } => args.hash(state),
+            _ => {
+                panic!("TODO: implement hashing for more variants.");
+            },
+        }
+    }
+}
+
+impl PartialEq for VisibleInst {
+    fn eq(&self, other: &VisibleInst) -> bool {
+        if self.0.opcode() != other.0.opcode() {
+            return false;
+        }
+
+        // TODO: Surely there is a better way to do this? Should we provide a
+        // PartialEq implementation for InstructionData? Note that it isn't as
+        // simple as just using the default implementation, since
+        // InstructionData contains things like floating-point literals which
+        // don't have Eq defined.
```

Comment:
Right now, those floating point literals are represented as `Ieee32` and `Ieee64` structs which contain floats. I have been wanting to change the internal representation to `u32`/`u64` instead. That would also enable you to `#[derive(PartialEq, Eq)]`, with the same caveat about value lists.

We don't want GVN to merge +0.0 and -0.0 either.